### PR TITLE
Add mysql example

### PR DESCRIPTION
**Pull request by oshai**
The example is still not working because of the following error:
`java.lang.IllegalStateException: Unable to create a ConnectionFactory for 'ConnectionFactoryOptions{options={user=test, driver=mysql, host=localhost, password=REDACTED, port=32782}}'. Available drivers: [ postgresql, h2, mssql ]
`

Probably need to include changes from mysql branch of spring-data-r2dbc:
https://github.com/spring-projects/spring-data-r2dbc/compare/mysql

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@oshai Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=r2dbc/r2dbc-client&pullRequestId=46)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=r2dbc/r2dbc-client&pullRequestId=46) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@oshai Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=r2dbc/r2dbc-client&pullRequestId=46)!

**Comment by oshai** [comment]
I added some fixes. one test pass and the rest fails. It looks like related to parsing of parameters for the prepared statement. I can fix it but want to be sure what is the expected behaviour:
```
java.lang.AssertionError: expectation "values from insertions" failed (expected: count = 10; actual: counted = 0; signal: onError(java.lang.NumberFormatException: For input string: "P0"))

	at reactor.test.ErrorFormatter.assertionError(ErrorFormatter.java:105)
	at reactor.test.ErrorFormatter.failPrefix(ErrorFormatter.java:94)
	at reactor.test.ErrorFormatter.fail(ErrorFormatter.java:64)
	at reactor.test.ErrorFormatter.failOptional(ErrorFormatter.java:79)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.checkCountMismatch(DefaultStepVerifierBuilder.java:1258)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onSignalCount(DefaultStepVerifierBuilder.java:1489)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onExpectation(DefaultStepVerifierBuilder.java:1341)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onError(DefaultStepVerifierBuilder.java:1030)
	at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onNext(MonoFlatMapMany.java:157)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1476)
	at reactor.core.publisher.MonoCompletionStage.lambda$subscribe$0(MonoCompletionStage.java:82)
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at com.github.jasync.sql.db.util.FutureUtilsKt.success(FutureUtils.kt:17)
	at com.github.jasync.sql.db.mysql.MySQLConnection.onOk(MySQLConnection.kt:184)
	at com.github.jasync.sql.db.mysql.codec.MySQLConnectionHandler.channelRead0(MySQLConnectionHandler.kt:119)
	at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:105)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:337)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1408)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:359)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:345)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:930)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:677)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:612)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:529)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:491)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:905)
	at java.lang.Thread.run(Thread.java:748)
	Suppressed: java.lang.NumberFormatException: For input string: "P0"
		at java.lang.NumberFormatException.forInputString(NumberFormatException.java:65)
		at java.lang.Integer.parseInt(Integer.java:580)
		at java.lang.Integer.parseInt(Integer.java:615)
		at com.github.jasync.r2dbc.mysql.JasyncStatement.bind(JasyncStatement.kt:44)
		at io.r2dbc.client.Update.bind(Update.java:57)
		at io.r2dbc.client.Example.lambda$null$4(Example.java:127)
		at java.util.stream.Streams$RangeIntSpliterator.forEachRemaining(Streams.java:110)
		at java.util.stream.IntPipeline$Head.forEach(IntPipeline.java:557)
		at io.r2dbc.client.Example.lambda$prepareStatement$5(Example.java:127)
		at io.r2dbc.client.R2dbc.lambda$withHandle$2(R2dbc.java:123)
		at reactor.core.publisher.MonoFlatMapMany$FlatMapManyMain.onNext(MonoFlatMapMany.java:153)
		... 31 more

```
@mp911de - can you assist?

**Comment by oshai** [comment]
to be more clear: the driver expect number s as index event if it is a string and gets `P0`. should the initial `P` be stripped?

File: src/test/java/io/r2dbc/client/MysqlExample.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            .option(PORT, SERVER.getPort())
+            .option(PASSWORD, SERVER.getPassword())
+            .option(USER, SERVER.getUsername())
+            .option(DATABASE, SERVER.getDatabase())
+            .build()));
+
+
+    @Override
+    public String getIdentifier(int index) {
+        return String.format("P%d", index);
```
Comment (mp911de) [inline src/test/java/io/r2dbc/client/MysqlExample.java]: This should be just a plain `int` to `String` conversion. The identifier is used with `bind(identifier, …)`.

File: src/test/java/io/r2dbc/client/MysqlExample.java [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        if (jdbcOperations == null) {
+            throw new IllegalStateException("JdbcOperations not yet initialized");
+        }
+
+        return jdbcOperations;
+    }
+
+    @Override
+    public String getPlaceholder(int index) {
```
Comment (mp911de) [inline src/test/java/io/r2dbc/client/MysqlExample.java]: This should be just `?`. The placeholder is used to construct the SQL statement and MySQL placeholders are `?`.

File: src/test/java/io/r2dbc/client/MysqlExample.java (hunk 1 above)
Comment (oshai) [reply src/test/java/io/r2dbc/client/MysqlExample.java]: got it.

File: src/test/java/io/r2dbc/client/MysqlExample.java (hunk 2 above)
Comment (oshai) [reply src/test/java/io/r2dbc/client/MysqlExample.java]: ok.

**Comment by oshai** [comment]
I added the fixes. compoundStatement test still fails, but I think that is not supported.

**Comment by mp911de** [comment]
Thanks a lot. I do not fully understand whether compound statements are a JDBC driver feature or whether newer MySQL versions support compound statements. 

I suggest disabling the compound statement test (`@Ignore`) with mentioning that compound statements are not supported.

**Comment by oshai** [comment]
It is supported by mysql server: https://dev.mysql.com/doc/internals/en/multi-statement.html but the driver do not support it at the moment. So I will disable the test

**Pull request closed without merging**